File: examples/python/api_examples.py
"""ProxyPay batch, webhook, and subscription examples. Python 3.10+."""

import hashlib
import hmac
import json
import os
import time
import urllib.request

BASE_URL = os.getenv("PROXYPAY_URL", "http://localhost:3000")


def request(method: str, path: str, token: str | None = None, body: bytes | None = None, content_type: str | None = None):
    headers = {"Authorization": f"Bearer {token}"} if token else {}
    if content_type:
        headers["Content-Type"] = content_type
    req = urllib.request.Request(BASE_URL + path, data=body, headers=headers, method=method)
    with urllib.request.urlopen(req) as response:
        return json.loads(response.read())


def upload_batch(csv_text: str, token: str):
    boundary = "proxypay-example-boundary"
    multipart = (f"--{boundary}\r\nContent-Disposition: form-data; name=\"file\"; filename=\"transactions.csv\"\r\nContent-Type: text/csv\r\n\r\n{csv_text}\r\n--{boundary}--\r\n").encode()
    job = request("POST", "/api/transactions/bulk", token, multipart, f"multipart/form-data; boundary={boundary}")
    while True:
        status = request("GET", f"/api/transactions/bulk/{job['jobId']}/status")
        if status["status"] in ("completed", "failed"):
            return status
        time.sleep(0.5)


def verify_webhook(raw_body: bytes, signature: str | None, secret: str) -> bool:
    if not signature or not signature.startswith("sha256="):
        return False
    expected = hmac.new(secret.encode(), raw_body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(signature[7:], expected)


def create_subscription(token: str):
    payload = json.dumps({"amount": "25.00", "currency": "USD", "interval": "monthly", "phone_number": "+237670000000"}).encode()
    return request("POST", "/api/subscriptions", token, payload, "application/json")["subscription"]


def pause_subscription(subscription_id: str, token: str):
    return request("POST", f"/api/subscriptions/{subscription_id}/pause", token)


if __name__ == "__main__":
    token = os.environ["PROXYPAY_TOKEN"]
    csv = "amount,phoneNumber,provider,stellarAddress\n10,+237670000000,MTN,GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWHF\n"
    print(upload_batch(csv, token))
    subscription = create_subscription(token)
    print(pause_subscription(subscription["id"], token))

File: examples/typescript/api-examples.ts
import { createHmac, timingSafeEqual } from "node:crypto";

const baseUrl = process.env.PROXYPAY_URL ?? "http://localhost:3000";
const token = process.env.PROXYPAY_TOKEN;

function authHeaders(): HeadersInit {
  if (!token) throw new Error("Set PROXYPAY_TOKEN before calling the API");
  return { Authorization: `Bearer ${token}` };
}

export async function uploadBatch(csv: string) {
  const form = new FormData();
  form.append("file", new Blob([csv], { type: "text/csv" }), "transactions.csv");
  const response = await fetch(`${baseUrl}/api/transactions/bulk`, {
    method: "POST",
    headers: authHeaders(),
    body: form,
  });
  if (!response.ok) throw new Error(`Batch upload failed: ${response.status} ${await response.text()}`);
  const job = (await response.json()) as { jobId: string };
  let status: { status: string; progress: { processed: number; total: number } };
  do {
    await new Promise((resolve) => setTimeout(resolve, 500));
    const poll = await fetch(`${baseUrl}/api/transactions/bulk/${job.jobId}/status`);
    if (!poll.ok) throw new Error(`Batch poll failed: ${poll.status}`);
    status = (await poll.json()) as typeof status;
  } while (status.status !== "completed" && status.status !== "failed");
  return { jobId: job.jobId, ...status };
}

export function verifyWebhook(rawBody: string, signature: string | undefined, secret: string): boolean {
  if (!signature?.startsWith("sha256=")) return false;
  const expected = createHmac("sha256", secret).update(rawBody).digest("hex");
  const received = Buffer.from(signature.slice(7), "utf8");
  const computed = Buffer.from(expected, "utf8");
  return received.length === computed.length && timingSafeEqual(received, computed);
}

export async function createSubscription() {
  const response = await fetch(`${baseUrl}/api/subscriptions`, {
    method: "POST",
    headers: { ...authHeaders(), "Content-Type": "application/json" },
    body: JSON.stringify({ amount: "25.00", currency: "USD", interval: "monthly", phone_number: "+237670000000" }),
  });
  if (!response.ok) throw new Error(`Subscription creation failed: ${response.status} ${await response.text()}`);
  return (await response.json()) as { subscription: { id: string; status: string } };
}

export async function pauseSubscription(id: string) {
  const response = await fetch(`${baseUrl}/api/subscriptions/${id}/pause`, { method: "POST", headers: authHeaders() });
  if (!response.ok) throw new Error(`Subscription pause failed: ${response.status} ${await response.text()}`);
  return response.json() as Promise<{ paused: boolean }>;
}

if (process.argv[1]?.endsWith("api-examples.ts")) {
  const result = await uploadBatch("amount,phoneNumber,provider,stellarAddress\n10,+237670000000,MTN,GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWHF\n");
  console.log(result);
  const subscription = await createSubscription();
  console.log(await pauseSubscription(subscription.subscription.id));
}
